api/envoy/extensions/filters/http/rate_limit_quota/v3/rate_limit_quota.proto

```diff
@@ -0,0 +1,416 @@
+      // the first quota assignment.
+      // [#not-implemented-hide:]
+      BlockingBehavior block_requests = 2;
+    }
+  }
+
+  // Specifies the behavior when the bucket's assignment has expired, and can't be refreshed
+  // due to the RLQS server's unavailability.
+  message ExpiredAssignmentBehavior {
+    // Reuse the last known quota assignment, effectively extending it for  the duration
+    // specified in the :ref:`expired_assignment_behavior_timeout
+    // <envoy_v3_api_field_extensions.filters.http.rate_limit_quota.v3.RateLimitQuotaBucketSettings.ExpiredAssignmentBehavior.expired_assignment_behavior_timeout>`
+    // field.
+    message ReuseLastAssignment {
+    }
+
+    // Limit the time :ref:`ExpiredAssignmentBehavior
+    // <envoy_v3_api_msg_extensions.filters.http.rate_limit_quota.v3.RateLimitQuotaBucketSettings.ExpiredAssignmentBehavior>`
+    // is applied. If the server doesn't respond within this duration:
+    //
+    // 1. Selected ``ExpiredAssignmentBehavior`` is no longer applied.
+    // 2. The assignment is abandoned: the bucket is purged from the memory and not tracked until
+    //    a new request matched into the bucket.
+    // 3. Until the server returns a new assignment, :ref:`no_assignment_behavior
+    //    <envoy_v3_api_field_extensions.filters.http.rate_limit_quota.v3.RateLimitQuotaBucketSettings.no_assignment_behavior>`
+    //    configuration takes over.
+    //
+    // When the field is not set, the ``ExpiredAssignmentBehavior`` time is **not limited**:
+    // it's applied until the server returns a new assignment.
+    //
+    // .. attention::
+    //   Zero-second ``Duration`` expires **immediately**, while unset ``Duration``
+    //   **has no expiration date**.
+    google.protobuf.Duration expired_assignment_behavior_timeout = 1
+        [(validate.rules).duration = {gte {}}];
+
+    oneof expired_assignment_behavior {
+      option (validate.required) = true;
+
+      // Apply the rate limiting strategy to all requests matched into the bucket until the RLQS
+      // server returns the new assignment, or :ref:`expired_assignment_behavior_timeout
+      // <envoy_v3_api_field_extensions.filters.http.rate_limit_quota.v3.RateLimitQuotaBucketSettings.ExpiredAssignmentBehavior.expired_assignment_behavior_timeout>`
+      // runs out.
+      type.v3.RateLimitStrategy fallback_rate_limit = 2;
+
+      // Reuse the last known quota assignment until the RLQS server returns the new assignment, or
+      // :ref:`expired_assignment_behavior_timeout <envoy_v3_api_field_extensions.filters.http.rate_limit_quota.v3.RateLimitQuotaBucketSettings.ExpiredAssignmentBehavior.expired_assignment_behavior_timeout>`
+      // runs out.
+      ReuseLastAssignment reuse_last_assignment = 3;
+    }
+  }
+
+  // Customize the deny response to the requests over the rate limit.
+  message DenyResponseSettings {
+    // HTTP response code for denying non-gRPC HTTP requests.
```

Comment:
It determines it from the content-type header. Here is the [algorithm](https://github.com/envoyproxy/envoy/blob/7e24b5dcc61deea9ae482548ad7a03e7b5ab1b1f/source/common/grpc/common.cc#L34)

Reply:
Okay, that seems like a fine approach.  But let's document that here in the comments for other implementations.

Reply:
Addressed in https://github.com/envoyproxy/envoy/pull/19793/commits/8799cceb8feaa12c54459e8b6bb68f713f4fa7d4.

Reply:
I realized that I missed Yan's comment. IMO `oneof` works better. We still can apply the algorithm to select default deny message automatically, when `DenyResponseSettings deny_response_settings` is not set. I'll make this change to protos.

Reply:
Removed from `deny_response_settings`, and added a note on the default behavior when unset (explaining content-type used for determining grpc vs http): https://github.com/envoyproxy/envoy/pull/19793/commits/4b9d187f1772a27be3f39f01125099a9acb722ff

Reply:
Given that Envoy knows whether a given request is HTTP or gRPC and can choose accordingly, I think the `oneof` actually makes the config less flexible, because it requires duplicating the entire config in two different routes if you want the config to differ only by having a custom deny response for both HTTP and gRPC.

I think we should back out this change and go back to the way this was before.

Reply:
Reverted the `oneof` change. Then added a comment on how gRPC requests are determined in https://github.com/envoyproxy/envoy/pull/19793/commits/2bccd556535122c00369f304df1784ce19a8cf5e.